### PR TITLE
[WIP] Add e2e ci for kubernetes with kind.

### PR DESCRIPTION
**Pull request by liangyuanpeng**
### what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

Make sure xline can work with kubernetes, so add e2e ci for kubernetes.

### what changes does this pull request make?

 add e2e ci for kubernetes.

---
Now, it's sometimes fail, sometimes success.

successed: https://github.com/liangyuanpeng/Xline/actions/runs/5688707901
failed:  https://github.com/liangyuanpeng/Xline/actions/runs/5688679640/job/15418920092

### CODE REVIEW

**Comment by liangyuanpeng** [comment]
> Now, it's sometimes fail, sometimes success.

I think it have some problem on the Xline side and need to fix.

**Comment by Phoenix500526** [comment]
Thank you for your contribution. Please add a sign-off in the commit message and modify It to comply with the conventional commit format. For your information: [Xline commit message guidelines](https://github.com/xline-kv/Xline/blob/master/CONTRIBUTING.md#commit-message-guidelines)

**Comment by liangyuanpeng** [comment]
@Phoenix500526

 I will rebase it when this pr is ready for review.
For now, it's pending for here:
> I think it have some problem on the Xline side and need to fix.

It's need the Xline maintainer to help check ,Thanks.

**Comment by liangyuanpeng** [comment]
Seems like it's related:
- https://github.com/xline-kv/Xline/issues/387

let me check it.

**Comment by liangyuanpeng** [comment]
It's always successed if i disable xline debug log.

https://github.com/liangyuanpeng/Xline/actions/workflows/kube.yml

image

**Comment by iGxnon** [comment]
> It's always successed if i disabled xline debug log.
> 
> https://github.com/liangyuanpeng/Xline/actions/workflows/kube.yml
> 
> image

How about adding a script to display the debug log of xline within the container? You could consider adding a simple test in the CI that sends a basic request to xline and check the outputs. This might help us in analyzing any potential issues related to xline.

**Comment by liangyuanpeng** [comment]
> How about adding a script to display the debug log of xline within the container? You could consider adding a simple test in the CI that sends a basic request to xline and check the outputs. This might help us in analyzing any potential issues related to xline.

Absolutely, I just cut the log here, seems like the xline server have some problem. FYI

grep `conflict`
```shell
2023-08-08T08:01:16.533848Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: ASReady(ProposeId("node1-657da3b8-0442-495a-b674-74a4390bfc2d"), 486)
2023-08-08T08:01:16.949130Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: SpecExeReady(ProposeId("node1-1a24f9fa-e910-4a97-b7d1-eafad8928c32"), 487)
2023-08-08T08:01:16.949162Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: ASReady(ProposeId("node1-1a24f9fa-e910-4a97-b7d1-eafad8928c32"), 487)
2023-08-08T08:01:18.964632Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: SpecExeReady(ProposeId("node1-89e0ca86-814f-48f5-ba02-89a411b2f1c3"), 488)
2023-08-08T08:01:18.964658Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: ASReady(ProposeId("node1-89e0ca86-814f-48f5-ba02-89a411b2f1c3"), 488)
2023-08-08T08:01:20.500827Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: SpecExeReady(ProposeId("node1-10d77a7b-f4f2-4a19-958d-bcdb6fc0e683"), 489)
2023-08-08T08:01:20.500879Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: ASReady(ProposeId("node1-10d77a7b-f4f2-4a19-958d-bcdb6fc0e683"), 489)
2023-08-08T08:01:20.592684Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: SpecExeReady(ProposeId("node1-6b5111f6-628c-4cb1-b56c-12c59afd41e7"), 490)
2023-08-08T08:01:20.592766Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: ASReady(ProposeId("node1-6b5111f6-628c-4cb1-b56c-12c59afd41e7"), 490)
2023-08-08T08:01:20.978428Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: SpecExeReady(ProposeId("node1-0d4ea7a2-fb5c-4818-b797-1db8b7774564"), 491)
2023-08-08T08:01:20.978449Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: ASReady(ProposeId("node1-0d4ea7a2-fb5c-4818-b797-1db8b7774564"), 491)
2023-08-08T08:01:22.992134Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: SpecExeReady(ProposeId("node1-bb7c75ac-e5bc-4d77-87eb-dd213dc86340"), 492)
2023-08-08T08:01:22.992162Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: ASReady(ProposeId("node1-bb7c75ac-e5bc-4d77-87eb-dd213dc86340"), 492)
2023-08-08T08:01:24.247018Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: SpecExeReady(ProposeId("node1-7ef08fbe-ac08-4fb2-b1e6-7f5135aad687"), 493)
2023-08-08T08:01:24.247057Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: ASReady(ProposeId("node1-7ef08fbe-ac08-4fb2-b1e6-7f5135aad687"), 493)
2023-08-08T08:01:25.006454Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: SpecExeReady(ProposeId("node1-c3c6aa14-c3e3-41c5-95e1-43181ade0704"), 494)
2023-08-08T08:01:25.006476Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: ASReady(ProposeId("node1-c3c6aa14-c3e3-41c5-95e1-43181ade0704"), 494)
2023-08-08T08:01:27.019814Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: SpecExeReady(ProposeId("node1-58be4d25-ae1d-4a5c-bc9e-81472bc3baab"), 495)
2023-08-08T08:01:27.019862Z DEBUG curp::server::cmd_worker::conflict_checked_mpmc: new ce event: ASReady(ProposeId("node1-58be4d25-ae1d-4a5c-bc9e-81472bc3baab"), 495)
```

grep `error`
```shell
ror(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.829271Z DEBUG tonic::codec::decode: decoder inner stream error: Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.829287Z  WARN xline::server::watch_server: Receive WatchRequest error Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.829395Z DEBUG tonic::codec::decode: decoder inner stream error: Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.829415Z  WARN xline::server::watch_server: Receive WatchRequest error Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.829498Z DEBUG tonic::codec::decode: decoder inner stream error: Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.829519Z  WARN xline::server::watch_server: Receive WatchRequest error Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.829540Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829557Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829566Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829573Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829580Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829650Z DEBUG tonic::codec::decode: decoder inner stream error: Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.829668Z  WARN xline::server::watch_server: Receive WatchRequest error Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.829688Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829703Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829718Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829730Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829816Z DEBUG tonic::codec::decode: decoder inner stream error: Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.829837Z  WARN xline::server::watch_server: Receive WatchRequest error Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.829858Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829873Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829886Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829898Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829912Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829924Z DEBUG hyper::proto::h2::server: stream error: error writing a body to connection: send stream capacity unexpectedly closed
2023-08-08T08:01:27.829988Z  WARN xline::server::watch_server: Receive WatchRequest error Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.830058Z  WARN xline::server::watch_server: Receive WatchRequest error Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
2023-08-08T08:01:27.829293Z  WARN xline::server::watch_server: Receive WatchRequest error Status { code: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }
```


The log generate from `v0.5.0`

This issue https://github.com/xline-kv/Xline/issues/402 also have some error, but is not same with here.

**Comment by liangyuanpeng** [comment]
Releated: https://github.com/xline-kv/Xline/issues/677

File: .github/workflows/kube.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      - name: chmod
+        if: always()
+        run: |
+          sudo chmod -R 777 /xline/kind
+          ls /xline/kind
+          ls /xline/kind/pods
+          du -h --max-depth=1 /xline/kind
+          ls -allh /xline/kind/containers
+      - name: upload kind logs
+        if: always()
```
Comment (Phoenix500526) [inline .github/workflows/kube.yml:42]: I think uploading detailed logs when deployment failed is a better choice.

**Review by Phoenix500526 — COMMENTED** [review]
Please remove trailing spaces and squash these commits.

**Comment by liangyuanpeng** [comment]
Create a clean PR https://github.com/xline-kv/Xline/pull/696 to work for it, closing..

**Pull request closed without merging**